File: rtttl-android/app/src/main/java/com/dgmltn/doorbell/MainActivity.java
package com.dgmltn.doorbell;

import android.media.AudioAttributes;
import android.media.MediaDataSource;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Toast;

import butterknife.ButterKnife;
import butterknife.OnClick;

import static android.media.AudioAttributes.CONTENT_TYPE_MUSIC;
import static android.media.AudioAttributes.USAGE_MEDIA;

public class MainActivity extends AppCompatActivity {

    MediaPlayer mediaPlayer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
    }

    @OnClick(R.id.start)
    public void start(View view) {
        if (mediaPlayer != null) {
            mediaPlayer.release();
        }

//		mediaPlayer = MediaPlayer.create(getApplicationContext(), R.raw.halloween);
        mediaPlayer = new MediaPlayer();

        AudioAttributes attributes = new AudioAttributes
            .Builder()
            .setContentType(CONTENT_TYPE_MUSIC)
            .setUsage(USAGE_MEDIA)
            .build();
        mediaPlayer.setAudioAttributes(attributes);

        //TODO: fix flutter plugin to use setAudioAttributes
//			mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);

        MediaDataSource source = new StringMediaDataSource(
            "Bod:d=8,o=5,b=250:a,b,a,4c#6,a,4c#6,c#6,4b,16a,16b,16a,e,4p,e,f#,e,4g,4g,4g,4f#,e," +
                "2p,a,b,a,4c#6,a," +
                "4c#6,c#6,4b,16a,16b,16a,e,4p,e,f#,e,4g,4g,4g,4g,g#,a,b,4e,4p,a,c#6,a,p,f#,e,4p,a,c#6,a,p,f#,e,2p,4f#,4a,4a,4b,4c6,b,a,p,a,b,a," +
				"4c#6,a,4c#6,c#6,4b,16a,16b,16a,e,4p,e,f#,e,4g,4g,4g,4g#,a");
        mediaPlayer.setDataSource(source);
        mediaPlayer.setOnCompletionListener(onCompletion);
        mediaPlayer.start();
    }

    @OnClick(R.id.stop)
    public void stop(View view) {
        if (mediaPlayer != null) {
            mediaPlayer.stop();
            mediaPlayer.release();
        }
    }

    MediaPlayer.OnCompletionListener onCompletion = new MediaPlayer.OnCompletionListener() {
        @Override
        public void onCompletion(MediaPlayer mp) {
            mediaPlayer.release();
            Toast.makeText(getApplicationContext(),
                           "onCompletion",
                           Toast.LENGTH_LONG)
                 .show();
        }
    };

    class StringMediaDataSource extends MediaDataSource {

        private final byte[] mSource;

        public StringMediaDataSource(String source) {
            mSource = source.getBytes();
        }

        @Override
        public int readAt(long position, byte[] buffer, int offset, int size) {

            int length = mSource.length;
            if (position >= length) {
                return -1; // -1 indicates EOF
            }

            if (position + size > length) {
                size -= (position + size) - length;
            }

            System.arraycopy(mSource, (int) position, buffer, offset, size);
            return size;
        }

        @Override
        public long getSize() {
            return mSource.length;
        }

        @Override
        public void close() {
        }
    }
}

File: tyson-android/app/src/main/java/com/dgmltn/tyson/MainActivity.kt
package com.dgmltn.tyson

import android.media.MediaPlayer
import android.os.AsyncTask
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.widget.Toast
import timber.log.Timber

class MainActivity : AppCompatActivity() {

    val recycler by lazy { findViewById(R.id.ringtones) as RecyclerView }
    val adapter by lazy { RingtonesAdapter(this) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        adapter.listener = { deviceSetRingtone(it.index) }
        recycler.layoutManager = LinearLayoutManager(this)
        recycler.adapter = adapter
    }

    override fun onStart() {
        super.onStart()

        AsyncTask.execute {
            val current = App.device.getRingtone()
            runOnUiThread { adapter.selectedIndex = current }
        }
    }

    ///////////////////////////////////////////////////////////////////////////////////////////////
    // Particle functions
    ///////////////////////////////////////////////////////////////////////////////////////////////

    private fun doBeepbeep() {
        AsyncTask.execute {
            App.device.beepbeep()
        }
    }

    private fun doDoorbell() {
        AsyncTask.execute {
            App.device.dingdong()
        }
    }

    private fun deviceSetRingtone(index: Int) {
        Timber.e("deviceSetRingtone($index)")
        AsyncTask.execute {
            App.device.setRingtone(index)
        }
    }

    ///////////////////////////////////////////////////////////////////////////////////////////////
    // RTTTL playback
    ///////////////////////////////////////////////////////////////////////////////////////////////

    private var mediaPlayer: MediaPlayer? = null

    private val onCompletion = MediaPlayer.OnCompletionListener {
        stop()
        Toast.makeText(this, "onCompletion", Toast.LENGTH_SHORT).show()
    }

    fun start(rawResId: Int) {
        stop()
        mediaPlayer = MediaPlayer.create(this, rawResId)
        if (mediaPlayer == null) {
            Toast.makeText(this, "unable to play", Toast.LENGTH_SHORT).show()
        } else {
            mediaPlayer?.setOnCompletionListener(onCompletion)
            mediaPlayer?.start()
        }
    }

    fun stop() {
        if (mediaPlayer != null) {
            if (mediaPlayer?.isPlaying ?: false) {
                mediaPlayer?.stop()
            }
            mediaPlayer?.release()
            mediaPlayer = null
        }
    }

}
